poc/vdaf_poc/field.py

```diff
@@ -272,6 +264,103 @@ def poly_eval(field: type[F], p: list[F], eval_at: F) -> F:
+    # all the ith divided differences. Third, we mutliply the ith basis
+    # polynomial by the one of the ith divided differences, and add the product
+    # to the output accumulator. Computation of both the basis polynomials and
+    # the divided differences are done recurrently, depending on values from
+    # just the previous iteration. However, we calculate a full triangle of
+    # divided difference values, one row per outer loop iteration, and only
+    # those along the left side (involving y_0) are directly used in the
+    # polynomial interpolation formula. The rest of the triangle is just
+    # computed to be used in subsequent rows of the triangle.
+    #
+    # Newton basis polynomials:
+    # n_i(x) = \prod_{j=0}^{i-1} (x - x_j)
+    # n_0(x) = 1
+    # n_i(x) = n_{i-1}(x) \cdot (x - x_{i - 1})
+    #
+    # Divided differences:
+    # [y_j] = y_j
+    # [y_j, y_{j+1}] = ([y_j] - [y_{j+1}]) / (x_j - x_{j+1})
+    # [y_j, ..., y_k] = ([y_j, ..., y_k] - [y_j, ..., y_k]) / (x_j - x_k)
+    #
+    # Newton polynomial interpolation:
+    # p(x) = \sum_{i=0}^{n-1} [y_0, ..., y_{i}] \cdot n_i(x)
+
+    # Handle i=0 as a special case via initialization of variables.
+    # First basis polynomial: n_0(x) = 1
+    previous_basis_polynomial: list[F] = [one]
+    # The top row of the triangle of divided differences is just every y-value.
+    previous_divided_differences = ys
+    # Initialize the output polynomial with the constant y0. (This is equal to
+    # [y0]*n_0(x) = y0 * 1)
+    output: list[F] = [ys[0]]
+
+    for i in range(1, n):
+        next_basis_polynomial = poly_mul(
+            field,
+            previous_basis_polynomial,
+            [-xs[i - 1], one],
+        )
+        previous_basis_polynomial = next_basis_polynomial
+
+        next_divided_differences: list[F] = []
+        for k in range(len(previous_divided_differences) - 1):
+            next_divided_differences.append(
+                (previous_divided_differences[k]
+                 - previous_divided_differences[k + 1])
+                * (xs[k] - xs[k + i]).inv()
+            )
+        previous_divided_differences = next_divided_differences
+
+        output = poly_add(
+            field,
+            output,
+            poly_mul(
+                field,
+                [next_divided_differences[0]],
+                next_basis_polynomial,
+            ),
+        )
+
+    return output
+
+
+def xgcd(a: int, b: int) -> tuple[int, int, int]:
+    """
+    Extended Euclidean algorithm.
+
+    Both a and b must be positive integers.
+    """
+    last_remainder, remainder = a, b
+    a, last_a, b, last_b = 0, 1, 1, 0
+    while remainder:
+        last_remainder, (quotient, remainder) = (
+            remainder,
+            divmod(last_remainder, remainder),
+        )
+        a, last_a = last_a - quotient * a, a
+        b, last_b = last_b - quotient * b, b
+    return last_remainder, last_a, last_b
```

Comment:
Suggestion: I generally prefer adding parentheses when constructing/destructing tuples. In any case I think we should try to be consistent. At the moment there's a mixture of both.
```suggestion
    return (last_remainder, last_a, last_b)
```